python/pyfeelpp-mor/feelpp/mor/nirb/nirb.py

```diff
@@ -0,0 +1,1282 @@
+            else:
+                raise ValueError("Unknown toolbox")
+
+
+    def initCoarseToolbox(self):
+        """Initialize the rectification problem
+        """
+        self.tbCoarse = self.setToolbox(self.H,mesh_path=self.coarsemesh_path)
+        self.XH = feelpp.functionSpace(mesh=self.tbCoarse.mesh(), order=self.order)
+        if self.worldcomm.isMasterRank():
+            print(f"[NIRB] Number of nodes on the coarse mesh : {self.tbCoarse.mesh().numGlobalPoints()}")
+
+
+    def setToolbox(self, hsize, mesh_path=None):
+        """Set up the toolbox object for the given model and mesh
+
+        Args:
+        -----
+            hsize (float): size of the mesh
+
+        Returns:
+        --------
+            Toolbox: toolbox object
+        """
+        if mesh_path == None: mesh_path=self.finemesh_path
+        # load meshes
+        mesh_ = feelpp.mesh(dim=self.dimension, realdim=self.dimension)
+        mesh = feelpp.load(mesh_, mesh_path, hsize)
+
+        # set mesh and model properties
+        if self.toolboxType == "heat":
+            tb = heat.heat(dim=self.dimension, order=self.order)
+        elif self.toolboxType == "fluid":
+            tb = fluid.fluid(dim=self.dimension)
+        else:
+            raise ValueError("Unknown toolbox")
+
+        tb.setMesh(mesh)
+        tb.setModelProperties(self.model)
+
+        tb.init()
+
+        return tb
+
+    def getField(self, toolbox):
+        """Get field of interest from the toolbox
+
+        Args:
+            toolbox (Toolbox): Tolbox object
+
+        Raises:
+            ValueError: Unknow toolbox
+
+        Returns:
+            feelpp_.discr.Element_*: field of the solution
+        """
+        if self.toolboxType == "heat":
+            return toolbox.fieldTemperature()
+        elif self.toolboxType == "fluid":
+            return toolbox.fieldVelocity()
+        else:
+            raise ValueError("Unknown toolbox")
+
+
+    def getToolboxSolution(self, tb, mu):
+        """Get the solution of the toolbox tb for the parameter mu
+
+        Args:
+            tb (Toolbox): Toolbox object
+            mu (parameterSpaceElement): parameter
+
+        Returns:
+            feelpp_.discr.Element_*: field of the solution
+        """
+        assembleToolbox(tb, mu)
+        tb.solve()
+        return self.getField(tb)
+
+
+    def createInterpolator(self, domain_tb, image_tb):
+        """Create an interpolator between two toolboxes
+
+        Args:
+            domain_tb (Toolbox): coarse toolbox
+            image_tb  (Toolbox): fine toolbox
+
+        Returns:
+            OperatorInterpolation: interpolator object
+        """
+        if self.toolboxType == "heat":
+            Vh_image = image_tb.spaceTemperature()
+            Vh_domain = domain_tb.spaceTemperature()
+        elif self.toolboxType == "fluid":
+            Vh_image = image_tb.spaceVelocity()
+            Vh_domain = domain_tb.spaceVelocity()
```

Comment:
this is not good I think to have this everywhere, if we have to extend this to other toolboxes

Reply:
A facade needs to be created : please follow this: https://www.geeksforgeeks.org/facade-method-python-design-patterns/

Reply:
we need to discuss this before doing any major changes to the code